src/runtime_src/doc/toc/xrt_native_apis.rst

```diff
@@ -520,222 +369,161 @@ Let us review the example below where the buffer is allocated for the kernel's f
 
-.. code::
-    
-    --------------------------
-    Instance:        foo_1
-    Base Address: 0x1800000
+The ``xrt::kernel`` class provides **overloaded operator ()** to execute the kernel with a comma-separated list of arguments.  
 
-    Argument:          a
-    Register Offset:   0x10
-    
 
+The above c++ code block is demonstrating 
+  
+- The kernel execution using the ``xrt::kernel()`` operator with the list of arguments that returns a ``xrt::run`` object. This is an asynchronous API and returns after submitting the task.    
+- The member function ``xrt::run::wait()`` is used to block the current thread until the current execution is finished. 
+- The member function ``xrt::run::set_arg()`` is used to set one or more kernel argument(s) before the next execution. In the example above, only the last (3rd) argument is changed.  
+- The member function ``xrt::run::start()`` is used to start the next kernel execution with new argument(s).   
 
-However, XRT also provides APIs to obtain the register offset for CU arguments. In the below example C API ``xrtKernelArgOffset`` is used to obtain offset of third argument of the CU ``foo:foo_1``.  
 
+Other kernel APIs
+~~~~~~~~~~~~~~~~~
 
-.. code:: c
-      :number-lines: 38
+**Obtaining the run object before execution**: Example of the previous section shows to obtain a ``xrt::run`` object when the kernel is executed (kernel execution returns a run object). However, a ``xrt::run`` object can be obtained even before the kernel execution. The flow is as below
 
-           // Assume foo has 3 arguments, a,b,c (arg 0, arg 1 and arg 2 respectively) 
-           
-           xrtKernelHandle kernel = xrtPLKernelOpenExclusive(device, xclbin_uuid, "foo:{foo_1}");
-           uint32_t arg_c_offset = xrtKernelArgOffset(kernel, 2);
- 
+- Open a Run object by the ``xrt::run`` constructor with a kernel argument). 
+- Set the kernel arguments associated for the next execution by the member function ``xrt::run::set_arg()``. 
+- Execute the kernel by the member function ``xrt::run::start()``.
+- Wait for the execution finish by the member function ``xrt::run::wait()``. 
 
-**C++**: The equivalent C++ API example
+**Timeout while wait for kernel finish**: The member function ``xrt::run::wait()`` blocks the current thread until the kernel execution finishes. To specify a timeout supported API ``xrt::run::wait()`` also accepts a timeout in millisecond unit.
 
-.. code:: c++
-      :number-lines: 38
+**Asynchronous update of the kernel arguments**: The member function ``xrt::run::set_arg()`` is synchronous to the kernel execution. This function can only be used when kernel is in the IDLE state and before the start of the next execution. An asynchronous version of this functionality (only for edge platform) is provided by the member function ``xrt::run::update_arg()`` to change the kernel arguments asynchronous to the kernel execution. 
```

Comment:
Uh oh, I wish we didn't document update_arg().   It is not a cool API.   Mailbox support in 21.2 is supposed to fix this.   update_arg is bad because the only way it writes word-by-word and the kernel is running, it could be using garbage until all words are written.   For example changing memory mapped arguments like this is scary, who knows where the kernel is reading or writing while the buffer address is written?

Reply:
removed